### PR TITLE
sql: pull password hashing back to the `sql` package

### PR DESCRIPTION
**Pull request by knz**
This also removes some obsolete code, and ensures that the username
validity check is performed before the password validity check.

Release note: None

### CODE REVIEW

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51501)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RichardJCai)

___
*[pkg/sql/alter_role.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51501#-MCNQDd50k5yAQiIrbBK:-MCNQDd50k5yAQiIrbBL:b-s5p4zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a920adfe72f44c547b175565ddbc24ec3250cfb8/pkg/sql/alter_role.go#L148)):*
> ```Go
> 		}
> 
> 		if hashedPassword == nil {
> ```

Nit: Does this logic for setting hashedPassword to an empty byte array make sense to move into checkPasswordAndGetHash? We can avoid duplication by doing so.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51501)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/alter_role.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51501#-MCNQDd50k5yAQiIrbBK:-MCQhuzp-EHZ5DUQIDCk:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a920adfe72f44c547b175565ddbc24ec3250cfb8/pkg/sql/alter_role.go#L148)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Nit: Does this logic for setting hashedPassword to an empty byte array make sense to move into checkPasswordAndGetHash? We can avoid duplication by doing so.
</blockquote></details>

Good idea! Done.

**Comment by knz** [comment]
the CI is flaking on  #51543

File: pkg/sql/create_role.go [hunk 1]
```diff
@@ -299,3 +296,26 @@ func (ua *userNameInfo) resolveUsername() (string, error) {
...
+	if hashedPassword == nil {
+		// v20.1 and below crash during authentication if they find a NULL value
+		// in system.users.hashedPassword. v20.2 and above handle this correctly,
+		// but we need to maintain mixed version compatibility for at least one
+		// release.
+		// TODO(nvanbenschoten): remove this for v21.1.
+		hashedPassword = []byte{}
+	}
+
+	return
```
Comment (RichardJCai) [inline pkg/sql/create_role.go:320]: Wait shouldn't this return hashedPassword or a return inside the if statement?

Comment (knz) [reply pkg/sql/create_role.go:320]: I don't understand your question. Can you rephrase?

Comment (RichardJCai) [reply pkg/sql/create_role.go:320]: Ah nvm I see that the return values are named, ignore my comment. 

Not a huge fan of named return values with an empty return statement though.
